client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,468 @@
+                                   HttpEntity entity, Header... headers) throws IOException {
+        URI uri = buildUri(endpoint, params);
+        HttpRequestBase request = createHttpRequest(method, uri, entity);
+        setHeaders(request, headers);
+        //we apply a soft margin so that e.g. if a request took 59 seconds and timeout is set to 60 we don't do another attempt
+        long retryTimeoutMillis = Math.round(this.maxRetryTimeoutMillis / (float)100 * 98);
+        IOException lastSeenException = null;
+        long startTime = System.nanoTime();
+        for (HttpHost host : nextHost()) {
+            if (lastSeenException != null) {
+                //in case we are retrying, check whether maxRetryTimeout has been reached, in which case an exception will be thrown
+                long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+                long timeout = retryTimeoutMillis - timeElapsedMillis;
+                if (timeout <= 0) {
+                    IOException retryTimeoutException = new IOException(
+                            "request retries exceeded max retry timeout [" + retryTimeoutMillis + "]");
+                    retryTimeoutException.addSuppressed(lastSeenException);
+                    throw retryTimeoutException;
+                }
+                //also reset the request to make it reusable for the next attempt
+                request.reset();
+            }
+
+            CloseableHttpResponse httpResponse;
+            try {
+                httpResponse = client.execute(host, request);
+            } catch(IOException e) {
+                RequestLogger.log(logger, "request failed", request, host, e);
+                onFailure(host);
+                lastSeenException = addSuppressedException(lastSeenException, e);
+                continue;
+            }
+            Response response = new Response(request.getRequestLine(), host, httpResponse);
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode < 300 || (request.getMethod().equals(HttpHead.METHOD_NAME) && statusCode == 404) ) {
+                RequestLogger.log(logger, "request succeeded", request, host, httpResponse);
+                onSuccess(host);
+                return response;
+            }
+            RequestLogger.log(logger, "request failed", request, host, httpResponse);
+            String responseBody;
+            try {
+                if (response.getEntity() == null) {
+                    responseBody = null;
+                } else {
+                    responseBody = EntityUtils.toString(response.getEntity());
+                }
+            } finally {
+                response.close();
+            }
+            lastSeenException = addSuppressedException(lastSeenException, new ResponseException(response, responseBody));
+            switch(statusCode) {
```

Comment:
There are two scenarios in our clients:
- request error: return error back, don't retry, call onSuccess
- node failure: retry all nodes, call onFailure at each failure (can happen with IOExceptions or 502, 503, 504).

500 is not seen as other 5xx codes like 502, 503 or 504, but like a request error. With 500 we do not retry, "connection" wise it's a success, we simply return the error and expect further action done by users. The assumption is that retrying won't fix the problem. This is what I am told all the language clients do, maybe somebody from @elastic/es-clients  would like to better explain this decision?